contractcourt/channel_arbitrator.go

```diff
@@ -1132,7 +1137,29 @@ func (c *ChannelArbitrator) shouldGoOnChain(htlcExpiry, broadcastDelta,
 
 	// We should on-chain for this HTLC, iff we're within out broadcast
 	// cutoff window.
-	return currentHeight >= broadcastCutOff
+	if currentHeight < broadcastCutOff {
+		return false
+	}
+
+	// In case of incoming htlc we should go to chain.
+	if htlc.Incoming {
+		return true
+	}
+
+	// For htlcs that are result of our initiated payments we give some grace
+	// period before force closing the channel. During this time we expect
+	// both nodes to connect and give a chance to the other node to send its
+	// updates and cancel the htlc.
+	// This shouldn't add any security risk as there is no incoming htlc to
+	// fulfill at this case and the expectation is that when the channel is
+	// active the other node will send update_fail_htlc to remove the htlc
+	// without closing the channel. It is up to the user to force close the
+	// channel if the peer misbehaves and doesn't send the update_fail_htlc.
+	// It is useful when this node is most of the time not online and is
+	// likely to miss the time slot where the htlc may be cancelled.
+	isForwarded := c.cfg.IsForwardedHTLC(c.cfg.ShortChanID, htlc.HtlcIndex)
```

Comment:
Is there a possibility of any unexpected behavior because of concurrency / race conditions? That the switch already modified the circuit when we get here?

Reply:
If he htlc is settled/cancelled just before this line is executed and therefore no longer exist in the switch the `isForwarded` will be `false` which will cause going on-chain after the grace period.
Such modification by the switch may as well happen before this code change and in that case  will cause going to chain immediately. 
I don't find this difference as problematic.

Reply:
Make sense. What happens in link and what happens in the arbitrator is indeed already desynchronized.